### PR TITLE
[feature](json-function) add json_insert, json_replace, json_set functions

### LINKED ISSUE #21268: [Enhancement] Improve Json function for doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now We have some _JSONB functions_ and  _JSON functions_ are not implement , specific semantics for each function below here can reference mysql

**NOTE** : _JSONB Functions_ , which has same function name with _JSON functions_ but with different implement which can be find in `be/src/vec/functions/function_jsonb.cpp ` . eg. json_parse alias jsonb_parse

### about json_path define:
https://modern-sql.com/blog/2017-06/whats-new-in-sql-2016#json-path

#### we have some JSONB functions which are not implement but JSON has already implement. below here
- [ ]  json_contains  @weijiew 
- [ ]  json_contains_path @weijiew 
- [ ]  json_quote json_unquote 
- [ ]  json_valid @hammerZh-Z 
- [ ]  json_array @yifei02 
- [ ]  json_object

#### below here we have both JSONB and JSON functions not implement .

- [ ]  ->>  ([->>](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_json-inline-path))
- [ ] JSON_DEPTH ([JSON_DEPTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-depth)) @niebayes 
- [ ] JSON_LENGTH ([JSON_LENGTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-length)) @Xuxiaotuan 
- [ ] JSON_KEYS ([JSON_KEYS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-keys)) @caicancai 
- [ ] JSON_SEARCH ([JSON_SEARCH()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-search)) 
- [ ] MERGE_OF ([MEMBER OF()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_member-of))

- [ ] JSON_INSERT ([JSON_INSERT()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-insert)).  @JasonWen1 
- [ ] JSON_REPLACE ([JSON_REPLACE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-replace)) @cjj2010 
- [ ] JSON_SET ([JSON_SET()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-set))

- [ ] JSON_OVERLAPS ([JSON_OVERLAPS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-overlaps))
- [ ] JSON_MERGE_PATCH ([JSON_MERGE_PATCH()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-patch))
- [ ] JSON_MERGE_PRESERVE ([JSON_MERGE_PRESERVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-preserve)) @taptao 
- [ ] JSON_REMOVE ([JSON_REMOVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-remove))
- [ ] JSON_TABLE ([JSON_TABLE()](https://dev.mysql.com/doc/refman/8.0/en/json-table-functions.html#function_json-table)) @lovlyhao 
- [ ] JSON_STORAGE_SIZE ([JSON_STORAGE_SIZE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-size))
- [ ] JSON_STORAGE_FREE ([JSON_STORAGE_FREE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-free))

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others

### Solution

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xuefengze**
## Proposed changes
Issue Number: close https://github.com/apache/doris/issues/21268

add json_insert, json_replace, json_set functions

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xuefengze** [comment]
run buildall

**Comment by xuefengze** [comment]
run buildall

**Comment by xuefengze** [comment]
run buildall

**Comment by xuefengze** [comment]
run p0

File: be/src/vec/functions/function_json.cpp [hunk 1]
```diff
@@ -1131,6 +1131,319 @@ class FunctionJsonUnquote : public IFunction {
     }
 };
 
+struct FunctionJsonInsert {
+    static constexpr auto name = "json_insert";
+    static constexpr auto is_insert = true;
```
Comment (eldenmoon) [inline be/src/vec/functions/function_json.cpp:1136]: why not use enum? boolean value is not intuitive

File: be/src/vec/functions/function_json.cpp [hunk 2]
```diff
@@ -1131,6 +1131,319 @@ class FunctionJsonUnquote : public IFunction {
...
+                std::vector<JsonPath> parsed_paths;
+
+#ifdef USE_LIBCPP
+                std::string s(path_string);
+                auto tok = get_json_token(s);
+#else
+                auto tok = get_json_token(path_string);
+#endif
+                std::vector<std::string> paths(tok.begin(), tok.end());
+                auto status = get_parsed_paths_with_status(paths, &parsed_paths);
```
Comment (eldenmoon) [inline be/src/vec/functions/function_json.cpp:1212]: why not use macro RETURN_IF_ERROR?

File: be/src/vec/functions/function_json.cpp [hunk 3]
```diff
@@ -1131,6 +1131,319 @@ class FunctionJsonUnquote : public IFunction {
...
+            std::string_view json_str(json_doc.data, json_doc.size);
+            objects[row].Parse(json_str.data(), json_str.size());
+            if (UNLIKELY(objects[row].HasParseError())) {
+                return Status::RuntimeError("invalid json doc function {}", get_name());
+            }
+        }
+
+        std::vector<std::vector<std::vector<JsonPath>>> json_paths;
+        auto status = get_parsed_path_columns(json_paths, data_columns, input_rows_count);
+        if (UNLIKELY(status != Status::OK())) {
```
Comment (eldenmoon) [inline be/src/vec/functions/function_json.cpp:1314]: use macro RETURN_IF_ERROR?

File: be/src/vec/functions/function_json.cpp (hunk 1 above)
Comment (xuefengze) [reply be/src/vec/functions/function_json.cpp:1136]: I will do it.

File: be/src/vec/functions/function_json.cpp (hunk 2 above)
Comment (xuefengze) [reply be/src/vec/functions/function_json.cpp:1212]: ok

**Comment by xuefengze** [comment]
run buildall

**Comment by xuefengze** [comment]
run p0

**Review by eldenmoon — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java [hunk 4]
```diff
@@ -424,6 +424,18 @@ protected FunctionCallExpr(FunctionCallExpr other) {
         this.isTableFnCall = other.isTableFnCall;
     }
 
+    public String parseJsonValueModifyDataType() {
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:427]: This function also need to be implemented in Nereids planner

Comment (xuefengze) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:427]: I'll implement it

**Comment by xuefengze** [comment]
run buildall

**Comment by xuefengze** [comment]
run p0

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by yiguolei** [comment]
@xuefengze  Hi, the code is conflict. Could you please resolve conflicts?

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
I have resolve the conflict and rebuild all

**Review by yiguolei — APPROVED** [review]
lgtm

**Review by eldenmoon — APPROVED** [review]
LGTM

**Pull request merged**